docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,179 @@
+# Kubernetes compute resource metrics API
+
+## Goals
+
+Provide resource usage metrics on pods and nodes on the API server to be used
+by the scheduler to improve job placement, utilization, etc. and by end users
+to understand the resource utilization of their jobs.
+
+## Current state
+
+Right now, the Kubelet exports container metrics via an API endpoint. This
+information is not gathered nor served by the Kubernetes API server.
+
+## Use cases
+
+The first user will be kubectl. The resource usage data can be shown to the
+user via a periodically refreshing interface similar to `top` on Unix-like
+systems. This info could let users assign resource limits more efficiently.
+
+```
+$ kubectl top kubernetes-minion-abcd
+Pod: monitoring-heapster-abcde
+CPU:
+    mean: 0.12 cores
+    95th: 1.10 cores
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
+	/api/v1/nodeMetrics/myNode/[derived|raw] 
```

Comment:
I'm fine with eric's arguments.  If Raw is fundamentally different than
Derived (or if I call one of them disproportionately often) then it seems
like one of them is /podmetrics and one of them is /derivedpodmetrics (if
they truly aren't the same resource)

On Wed, Jul 29, 2015 at 2:10 AM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/compute-resource-metrics-api.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35730196>
> :
>
> > +    mean: 0.12 cores
> > +    95th: 1.10 cores
> > +    max:  1.91 cores
> > +Memory:
> > +    mean: 300MB (10%)
> > +    95th: 330MB (11%)
> > +    max:  395MB (13%)
> > +```
> > +
> > +A second user will be the scheduler. To assign pods to nodes efficiently, the
> > +scheduler needs to know the current free resources on each node.
> > +
> > +## Proposed endpoints
> > +
> > +	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
> > +	/api/v1/nodeMetrics/myNode/[derived|raw]
>
> I did not want to get the metrics under
> /api/v1/namespaces/myns/pod/foo/metrics, for the reasons @erictune
> <https://github.com/erictune> mentioned. The data volume and rate of
> change are too high to include in the pod resource, and I didn't want to
> establish the precedent that a subresource could be used to access totally
> separate information.
>
> As for whether derived and raw should be subresources of a single
> resource, or whether they should be separate resources: Separate resources
> would make sense if we thought they wouldn't be generated and/or consumed
> together, which is probably the case, since most components will consume
> derived resources, which we expect to be produced at different timescales
> than the raw metrics.
>
> As for whether derived and raw appear to be subresources of pods and
> nodes, but in a different API group, or whether they are simply parallel
> resource structures: I prefer parallel resource structures, as we do for
> Endpoints. I have previously thought about nesting Endpoints under pods,
> replicationcontrollers, and services, to expose an API for
> ResourceLocation, but I'm no longer sure that's a good idea, for the
> reasons mentioned above.
>
> I could be potentially swayed by arguments on these points, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35730196>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift